### PR TITLE
Resume and Start chat buttons from landing page

### PR DESCRIPTION
**Pull request by noxware**
# Changes
- Replaces the "Downloaded..." button in the landing page with "Chat with Model" and "Resume".
  - These work the same as in the "My Models" screen.
- Tried to extract a common reusable button called `CButton` (to avoid name conflicts).
- Seems like a small issue with the cursor got fixed in the landing screen by using that button.

# Before
https://github.com/project-robius/moxin/assets/7684329/af64ed9d-1f7e-4e41-9c3b-1732e036581e

# After

https://github.com/project-robius/moxin/assets/7684329/fe195bac-465a-4275-96ab-145cf2b0ae63

https://github.com/project-robius/moxin/assets/7684329/ae417abe-3f57-406a-9f0c-eb0a1bc80b56

### CODE REVIEW

File: Cargo.lock [hunk 1]
```diff
@@ -2350,7 +2350,8 @@ dependencies = [
 [[package]]
```
Comment (noxware) [inline Cargo.lock:2350]: My rust-analyzer insists on updating the lock file so maybe it's not up to date? Is this update correct?

File: moxin-frontend/src/app.rs [hunk 2]
```diff
@@ -159,13 +159,7 @@ impl LiveRegister for App {
...
 
         // Shared
-        crate::shared::styles::live_design(cx);
-        crate::shared::resource_imports::live_design(cx);
-        crate::shared::widgets::live_design(cx);
-        crate::shared::icon::live_design(cx);
-        crate::shared::modal::live_design(cx);
-        crate::shared::external_link::live_design(cx);
-        crate::shared::popup::live_design(cx);
+        crate::shared::live_design(cx);
```
Comment (noxware) [inline moxin-frontend/src/app.rs:162]: wdyt about doing the same for landing, etc and removing the comments used for grouping?

File: moxin-frontend/src/data/store.rs [hunk 3]
```diff
@@ -9,7 +9,7 @@ use moxin_protocol::data::{
     DownloadedFile, File, FileID, Model, PendingDownload, PendingDownloadsStatus,
 };
 use moxin_protocol::protocol::{Command, LoadModelOptions, LoadModelResponse};
-use std::collections::{VecDeque, HashMap};
+use std::collections::{HashMap, VecDeque};
```
Comment (noxware) [inline moxin-frontend/src/data/store.rs:12]: These kind of changes are from Rust's formatter. I think is nice to keep them.

File: moxin-frontend/src/my_models/downloaded_files_table.rs [hunk 4]
```diff
@@ -271,6 +266,7 @@ impl Widget for DownloadedFilesTable {
     }
 
     fn draw_walk(&mut self, cx: &mut Cx2d, scope: &mut Scope, walk: Walk) -> DrawStep {
+        self.file_item_map.clear();
```
Comment (noxware) [inline moxin-frontend/src/my_models/downloaded_files_table.rs:269]: The hashmap I mentioned it could be leaking if it's not cleared each cycle :P

File: moxin-frontend/src/my_models/downloaded_files_table.rs [hunk 5]
```diff
@@ -544,59 +540,45 @@ pub enum ButtonType {
 #[derive(Live, LiveHook, Widget)]
 pub struct ActionButton {
     #[deref]
-    view: View,
+    deref: CButton,
```
Comment (fmzbl) [inline moxin-frontend/src/my_models/downloaded_files_table.rs:543]: This is a NIT, but we are using the name of the widget we deref everywhere else, so it itches me a bit.

File: moxin-frontend/src/shared/mod.rs [hunk 6]
```diff
@@ -6,3 +9,14 @@ pub mod resource_imports;
 pub mod styles;
 pub mod utils;
 pub mod widgets;
+
+pub fn live_design(cx: &mut Cx) {
```
Comment (fmzbl) [inline moxin-frontend/src/shared/mod.rs:13]: I would discuss this kind of changes first, and then apply them for everything if we wanna do them.

File: moxin-frontend/src/shared/actions.rs [hunk 7]
```diff
@@ -0,0 +1,9 @@
+use makepad_widgets::DefaultNone;
+use moxin_protocol::data::FileID;
+
+#[derive(Clone, DefaultNone, Debug)]
+pub enum DownloadedFileAction {
```
Comment (fmzbl) [inline moxin-frontend/src/shared/actions.rs:5]: since this action is now being dispatched and defined outside of the original widget file, i think we should find a new name for it, it became something else.

File: moxin-frontend/src/shared/widgets/c_button.rs [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+            icon_walk: {width: 12, height: 12}
+        }
+
+        label = <Label> {}
+    }
+);
+
+/// An attempt of a custom button, sharable with the rest of the app.
+#[derive(Widget, Live, LiveHook)]
+pub struct CButton {
```
Comment (fmzbl) [inline moxin-frontend/src/shared/widgets/c_button.rs:37]: Also, we are using views as buttons everywhere in the app, i think we should make this change in another pr, covering every case.

[the remark this answers is not in the document]
Comment (joulei) [reply moxin-frontend/src/shared/widgets/c_button.rs:37]: I'm also struggling to understand why CButton. 
Furthermore, I think I'd rather either keep a simple RoundedView (something simple quickly swappable for `Button`), or use Makepad's own `Button` already, I don't love that we're re-implementing a similar button over view, and adding similar but not consistent APIs (`tapped` vs `clicked`) is there something that prevents us from using `Button` today?

File: moxin-frontend/src/my_models/downloaded_files_table.rs [hunk 9]
```diff
@@ -544,59 +540,45 @@ pub enum ButtonType {
...
 }
 
 impl ActionButtonRef {
     pub fn set_visible(&mut self, visible: bool) {
         if let Some(mut inner) = self.borrow_mut() {
-            inner.view.visible = visible;
+            inner.deref.visible = visible;
         }
     }
 }
```
Comment (joulei) [inline moxin-frontend/src/my_models/downloaded_files_table.rs:580]: Our current convention is to use `inner.view...`instead of `deref`, IMO it's a bit more explicit so that when navigating code I can quickly see over what we're operating and I don't need to go check separately what we're derefing. I think `deref` is more of a technicality than semantics.

File: moxin-frontend/src/shared/widgets/c_button.rs [hunk 10]
```diff
@@ -0,0 +1,92 @@
+use makepad_widgets::*;
+
```
Comment (joulei) [inline moxin-frontend/src/shared/widgets/c_button.rs:2]: It took me a second to understand that `widgets.rs` has styling of existing widgets (no new widgets), and the new widgets folder has actual widgets.
I think we should either put new things like `c_button` in the root of `shared` module as we're already doing, or create a new submodule widgets (like you did now, and move all shared widgets there) but also rename `widgets.rs` or delete it and move its contents into styles.rs

File: moxin-frontend/src/shared/mod.rs (hunk 6 above)
Comment (joulei) [reply moxin-frontend/src/shared/mod.rs:13]: I think I brought this up before, it should be a good idea.
If we agree on the change, I wouldn't delete this now, but for next time I would do all of these in a separate PR.

File: moxin-frontend/src/my_models/downloaded_files_table.rs [hunk 11]
```diff
@@ -544,59 +540,45 @@ pub enum ButtonType {
...
 
     fn draw_walk(&mut self, cx: &mut Cx2d, scope: &mut Scope, walk: Walk) -> DrawStep {
-        self.view.draw_walk(cx, scope, walk)
+        self.deref.draw_walk(cx, scope, walk)
+    }
+}
+
+impl WidgetMatchEvent for ActionButton {
+    fn handle_actions(&mut self, cx: &mut Cx, actions: &Actions, scope: &mut Scope) {
+        if self.deref.tapped(actions) {
```
Comment (joulei) [inline moxin-frontend/src/my_models/downloaded_files_table.rs:561]: Same as the other comment about self.`view` vs self.`deref`.

File: moxin-frontend/src/data/store.rs [hunk 12]
```diff
@@ -49,6 +49,7 @@ pub struct DownloadInfo {
 pub struct ModelWithPendingDownloads {
     pub model: Model,
     pub pending_downloads: Vec<PendingDownload>,
+    pub current_file_id: Option<FileID>,
```
Comment (joulei) [inline moxin-frontend/src/data/store.rs:52]: I might be missing something but can you use the already existing `current_chat` in `Store` instead of adding this to `ModelWithPendingDownloads`? Since in `draw_walk` of `ModelFilesList` you have access to the Store. It might  look like more work on the other side but semantically it doesn't seem like `ModelWithPendingDownloads` would need a `current_file_id`.

**Review by joulei — COMMENTED** [review]
Appreciate you trying to find more ergonomic ways in Makepad.
However for this case I'd rather use Makepad's `Button` since it already has a `.clicked` API. I wouldn't be hesitant of having to create custom Widgets for lots of things if that means simplifying and shortening the code.
If the goal is to simplify and re-use, in this case it could be a lot simpler to make the `DownloadedFilesTable` rows their own Widget, and then just use `button.clicked()`. 

Don't necessarily take this a request of changes though, I'd be happy if we reach a middle ground were there's later a quick path to swap out these for Makepad's `Button` without much changes (see my comment about using RoundedView or matching API).

**Review by joulei — COMMENTED** [review]
I know we need new design for this, but having different colors in `Download` and `Resume Chat` feels a bit off to me. Could we perhaps just either use the same color in the meantime, or just the outline like in `Chat with Model`? (so that it doesn't look like we made a typo/mistake with the colors)

Screenshot 2024-05-24 at 9 26 09 AM

File: moxin-frontend/src/app.rs (hunk 2 above)
Comment (noxware) [reply moxin-frontend/src/app.rs:162]: Taking that as a yes

File: moxin-frontend/src/data/store.rs (hunk 12 above)
Comment (jmbejar) [reply moxin-frontend/src/data/store.rs:52]: I'm almost sure there is no way to reach the `store` from `ModelFilesList`. This is because we're passing more granular data at this widget (represented with the struct `ModelWithPendingDownloads`). So, I think @noxware was following the existing approach there.

File: moxin-frontend/src/landing/model_files_item.rs [hunk 13]
```diff
@@ -195,18 +215,33 @@ impl Widget for ModelFilesItem {
...
-
-                cx.widget_action(
-                    widget_uid,
-                    &scope.path,
-                    ModelFileItemAction::Download(file.clone(), model.clone()),
-                );
-            }
+        let widget_uid = self.widget_uid();
+
+        if self.cbutton(id!(download_button)).tapped(&actions) {
```
[the remark this answers is not in the document]
Comment (noxware) [reply moxin-frontend/src/landing/model_files_item.rs:220]: Yeah, I didn't want to imply a "mouse device" when this may be eventually be used in mobile so for me "tapped" or "[pressed](https://api.flutter.dev/flutter/material/ButtonStyleButton/onPressed.html)" are more "cross-platform" terms, but I agree that we should make everything consistent for now 👍

File: moxin-frontend/src/data/store.rs (hunk 12 above)
Comment (jmbejar) [reply moxin-frontend/src/data/store.rs:52]: Relevant code: https://github.com/project-robius/moxin/blob/main/moxin-frontend/src/landing/model_list.rs#L101-L103

(from this point deeper in the widgets tree, the scope does not hold the entire `store`)

File: moxin-frontend/src/landing/model_files_item.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (noxware) [reply moxin-frontend/src/landing/model_files_item.rs]: Change done

File: moxin-frontend/src/shared/mod.rs (hunk 6 above)
Comment (noxware) [reply moxin-frontend/src/shared/mod.rs:13]: I was reading the comments in order and since I saw @joulei thumbs up I applied the change for the rest of occurrences but I didn't saw this comment.

Shall I revert https://github.com/project-robius/moxin/pull/83/commits/d52c1ecdddedef424e33c380250685d36df8e469?

File: moxin-frontend/src/data/store.rs (hunk 12 above)
Comment (noxware) [reply moxin-frontend/src/data/store.rs:52]: Yes, what @jmbejar mentioned is the reason why I choose to do it this way. The `draw_walk` and `handle_event` receive scopes with different things. The draw walk receives what @jmbejar mentioned.

I think the store is accesible in the handle event (which surprised me and confused me a lot until I followed the code up to discover the origin of both scopes), but adding a `#rust` field and setting it on events as a workaround was worst in my opinion.

Extra note:
I noticed that makepad, apart of having a `Scope.with_data` has a `Scope.with_props`. The difference is that one is mutable and the later is immutable. If they are hold in different slots, then, would this problem be resolved by using `props` for `ModelWithPendingDownloads` and reserve `data` to always pass the global context (store) down?

File: moxin-frontend/src/shared/actions.rs (hunk 7 above)
Comment (joulei) [reply moxin-frontend/src/shared/actions.rs:5]: Oh yeah I meant that the Action itself is the same one. (And it's technically an action on a downloaded file either way). But I do like `ChatAction` 👍🏽, it's a bit more obvious

File: moxin-frontend/src/shared/widgets/c_button.rs (hunk 8 above)
Comment (noxware) [reply moxin-frontend/src/shared/widgets/c_button.rs:37]: Covered in slack discussion

**Review by joulei — APPROVED** [review]
